File: MATTHEW/leetcode/AIcoding/4.GPT-4o/javascript_solutions/210-course-schedule-ii.js
var findOrder = function(numCourses, prerequisites) {
    const graph = new Array(numCourses).fill(0).map(() => []);
    const inDegree = new Array(numCourses).fill(0);

    for (const [course, prereq] of prerequisites) {
        graph[prereq].push(course);
        inDegree[course]++;
    }

    const queue = [];
    for (let i = 0; i < numCourses; i++) {
        if (inDegree[i] === 0) {
            queue.push(i);
        }
    }

    const order = [];
    while (queue.length > 0) {
        const course = queue.shift();
        order.push(course);

        for (const nextCourse of graph[course]) {
            inDegree[nextCourse]--;
            if (inDegree[nextCourse] === 0) {
                queue.push(nextCourse);
            }
        }
    }

    return order.length === numCourses ? order : [];
};

File: leetcode/AIcoding/4.GPT-4o/go_solutions/210-course-schedule-ii.go
func findOrder(numCourses int, prerequisites [][]int) []int {
    indegree := make([]int, numCourses)
    graph := make([][]int, numCourses)

    for _, prereq := range prerequisites {
        graph[prereq[1]] = append(graph[prereq[1]], prereq[0])
        indegree[prereq[0]]++
    }

    queue := []int{}
    for i := 0; i < numCourses; i++ {
        if indegree[i] == 0 {
            queue = append(queue, i)
        }
    }

    order := []int{}
    for len(queue) > 0 {
        course := queue[0]
        queue = queue[1:]
        order = append(order, course)

        for _, nextCourse := range graph[course] {
            indegree[nextCourse]--
            if indegree[nextCourse] == 0 {
                queue = append(queue, nextCourse)
            }
        }
    }

    if len(order) == numCourses {
        return order
    }
    return []int{}
}